### PR TITLE
SOF IPC header cleanup

### PR DESCRIPTION
**Pull request by rfredzim**
Refactor of code related to IPC handling.
PR Includes:
- fix the parameters passed to ipc_msg_init
- update of common ipc4 structures to reflect primary/extension contents of ipc4 header
- update of ipc4 command specific structures to reflect primary/extension contents of ipc4 header
- remove the special treatment of the extension dword which had been part of payload data
- update the ipc_cmd_hdr to be used safely for ipc3 and ipc4

### CODE REVIEW

File: src/include/sof/ipc/common.h [hunk 1]
```diff
@@ -23,7 +23,9 @@ struct dma_sg_elem_array;
 struct ipc_msg;
 
 /* generic IPC header regardless of ABI MAJOR type that is always 4 byte aligned */
-typedef uint32_t ipc_cmd_hdr;
+typedef struct ipc_cmd_hdr_ {
+	uint32_t dat[2];
+} ipc_cmd_hdr;
```
Comment (lgirdwood) [inline src/include/sof/ipc/common.h:28]: This should be opaque and abstract - can we make this a generic pointer instead (and then cast appropriately in IPC3/4) ?

**Review by lgirdwood — COMMENTED** [review]
First 2 patches look good, but open on the generic ipc header type.

File: src/include/sof/ipc/common.h (hunk 1 above)
Comment (rfredzim) [reply src/include/sof/ipc/common.h]: Hi Liam,

The generic pointer (uint32_t *) as used previously is what I'm trying to fix in this patch among other issues.
There are spots in code that do the casting back and forth between this and the larger IPC structures and then access by indexing with respect to this pointer which is an unsafe practice. I'm trying to make the ipc solution more robust and error proof for the future by this.

BR, Rafal

Comment (lgirdwood) [reply src/include/sof/ipc/common.h]: Sorry, I mean we could do something generic/opaque like
```c
struct ipc_cmd_hdr; /* opaque type in header for declarations, defined in IPC3/4 handler code */
```

**Review by cujomalainey — APPROVED** [review]
LGTM for hotword changes

File: src/platform/intel/cavs/platform.c [hunk 2]
```diff
@@ -313,16 +312,15 @@ int platform_boot_complete(uint32_t boot_message)
 	mailbox_dspbox_write(0, &ready, sizeof(ready));
 
 	/* get any IPC specific boot message and optional data */
-	data = SRAM_WINDOW_HOST_OFFSET(0) >> 12;
-	ipc_boot_complete_msg(&header, &data);
+	ipc_boot_complete_msg(&header, SRAM_WINDOW_HOST_OFFSET(0) >> 12);
```
Comment (lyakh) [inline src/platform/intel/cavs/platform.c:315]: this seems strange - the second parameter of `ipc_boot_complete_msg()` isn't used now any more

**Review by lyakh — CHANGES_REQUESTED** [review]

File: src/platform/intel/cavs/platform.c (hunk 2 above)
Comment (lgirdwood) [reply src/platform/intel/cavs/platform.c:315]: @lyakh do you mean on CAVS what about other platforms ? We could make this a separate PR if never used ?

Comment (lyakh) [reply src/platform/intel/cavs/platform.c:315]: @lgirdwood I mean this very patch - it modifies `ipc_boot_complete_msg()` to not use its second parameter in both IPC3 and IPC4 implementations and still insists on putting some meaningfully looking data there. That seems strange to me. So, no, it cannot be a separate PR, it very much belongs to this PR

Comment (rfredzim) [reply src/platform/intel/cavs/platform.c:315]: @lyakh you're right previously the header part was set inside ipc_boot_complete_msg and data has been set outside of it, in most recent patchset I have corrected the the ipc_boot_complete_msg to keep the previous behavior for IPC3

File: src/include/sof/ipc/common.h (hunk 1 above)
Comment (rfredzim) [reply src/include/sof/ipc/common.h]: Hi @lgirdwood. I have done the changes to abstract the ipc_cmd_hdr for ipc3 / ipc4 according to your suggestion (had to do it in ipc*/header.h and not in ipc*/handler.c because the definition is also used outside of handler.c).
The thing is however that [checktree (IPC_VERSION=IPC4 UNSIGNED_list= SIGNED_list='tgl tgl-h cnl icl jsl')](https://github.com/thesofproject/sof/runs/6224422069?check_suite_focus=true#logs)
tries to build e.g. cavs/platform.c using ipc4.
Effectively in the tree we have a combination of a platform and different MAJOR ABI (e.g. cavs + IPC3 and cavs + IPC4) which brings me to conclusion that ipc_cmd_hdr definition should in fact be common for IPC3/IPC4 to keep the platform code happy.
Do you have other suggestion in this case?

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
@rfredzim can you check CI, thanks

File: src/include/ipc3/header.h [hunk 3]
```diff
@@ -0,0 +1,40 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright(c) 2021 Intel Corporation. All rights reserved.
+ */
```
[the remark this answers is not in the document]
Comment (lyakh) [reply src/include/ipc3/header.h:4]: I see, my mistake, thanks

File: src/ipc/ipc4/handler.c [hunk 4]
```diff
@@ -940,18 +940,14 @@ ipc_cmd_hdr *ipc_compact_read_msg(void)
...
 	msg_data.msg_out[0] = msg->header;
-	msg_data.msg_out[1] = *(uint32_t *)msg->tx_data;
+	msg_data.msg_out[1] = msg->extension;
 
-	/* the first uint of msg data is sent by ipc data register for ipc4 */
-	size = msg->tx_size - sizeof(uint32_t);
-	if (size)
-		mailbox_dspbox_write(0, (uint32_t *)msg->tx_data + 1, size);
+	if (msg->tx_size)
+		mailbox_dspbox_write(0, (uint32_t *)msg->tx_data, msg->tx_size);
```
Comment (kv2019i) [inline src/ipc/ipc4/handler.c:946]: I.e. this would benefit from some description in commit. Apparently the logic how to handle the special treatment of the first word, is changed in this patch. This is probably all good, but given title "header cleanup", the above code change jumps out quite a bit.

**Review by kv2019i — COMMENTED** [review]
This seems to be a bit more than header cleanup. :) Mostly looks good, a couple of questions inline. @ujfalusi  @ranj063 @RanderWang -- do we need to update kernel headers as well. 
@rfredzim I'm assuming this is ABI compatible, so kernel change is strictly not needed, but for change tracking better to have the data structures aligned.

File: src/audio/mixer.c [hunk 5]
```diff
@@ -746,7 +746,7 @@ static int mixin_bind(struct comp_dev *dev, void *data)
 
 	bu = (struct ipc4_module_bind_unbind *)data;
 	src_id = IPC4_COMP_ID(bu->header.r.module_id, bu->header.r.instance_id);
-	sink_id = IPC4_COMP_ID(bu->data.r.dst_module_id, bu->data.r.dst_instance_id);
+	sink_id = IPC4_COMP_ID(bu->header_ext.r.dst_module_id, bu->header_ext.r.dst_instance_id);
```
Comment (ujfalusi) [inline src/audio/mixer.c:749]: Or just simply `extension` ?

File: src/include/ipc4/module.h [hunk 6]
```diff
@@ -142,7 +142,7 @@ struct ipc4_module_init_instance {
 			uint32_t extended_init      : 1;
 			uint32_t _hw_reserved_2     : 2;
 		} r;
-	} data;
+	} header_ext;
```
Comment (ujfalusi) [inline src/include/ipc4/module.h:145]: I think in IPC4 it is more like that we have a `header` which is u64 and it has two u32 parts: `primary` and an `extension`.
The `header` (as two u32) is sent via registers back and forth between the host and DSP.
Additional data is via shared mailbox memory (set/get large).

File: src/include/sof/ipc/common.h (hunk 1 above)
Comment (lgirdwood) [reply src/include/sof/ipc/common.h]: > Effectively in the tree we have a combination of a platform and different MAJOR ABI (e.g. cavs + IPC3 and cavs + IPC4) which brings me to conclusion that ipc_cmd_hdr definition should in fact be common for IPC3/IPC4 to keep the platform code happy.
> Do you have other suggestion in this case?

So IIUC, we have a generic type for header that we define in IPC3 and IPC4 code which is fine. We could also define this in the IPC driver logic to align with the HW (and not the IPC flavour), meaning the IPC drivers will always build regardless of any IPC version (including any future IPC updates).

File: .github/workflows/installer.yml [hunk 7]
```diff
@@ -24,7 +24,6 @@ jobs:
       matrix:  # just a vector in this case
         make_env: [
           IPC_VERSION=,  # default version
-          IPC_VERSION=IPC4 UNSIGNED_list= SIGNED_list='tgl tgl-h cnl icl jsl',
```
Comment (lgirdwood) [inline .github/workflows/installer.yml:27]: Need to leave this in atm as CI and most developers only have CAVS platforms for test and dev work on IPC4.

**Review by lgirdwood — CHANGES_REQUESTED** [review]
LGTM, just need to keep the IPC4 support on the CAVS platforms for the time being.

File: .github/workflows/installer.yml (hunk 7 above)
Comment (marc-hb) [reply .github/workflows/installer.yml:27]: This removes IPC4 compilation for all platforms without replacing them with any IPC4 platform. So this is effectively turning off IPC4 compilation entirely which makes no sense and makes even less sense in an IPC4 pull request. This looks like trying to hide compilation issues.

As Liam mentioned, IPC4 testing is currently performed on IPC3 platforms. Even if it were not, this type of "poor man's randconfig" is useful because it can find logical Kconfig bugs. You can never have too much randconfig.

**Review by marc-hb — CHANGES_REQUESTED** [review]

**Comment by RanderWang** [comment]
/media/rander/linux_data/sof/src/ipc/ipc4/handler.c: In function ‘ipc4_delete_pipeline’:
/media/rander/linux_data/sof/src/ipc/ipc4/handler.c:70: error: ‘struct ipc4_pipeline_delete’ has no member named ‘header’
/media/rander/linux_data/sof/src/ipc/ipc4/handler.c:72: error: ‘struct ipc4_pipeline_delete’ has no member named ‘header’
cc1: warnings being treated as errors
/media/rander/linux_data/sof/src/ipc/ipc4/handler.c: In function ‘ipc_cmd’:
/media/rander/linux_data/sof/src/ipc/ipc4/handler.c:970: warning: initialization from incompatible pointer type
/media/rander/linux_data/sof/src/ipc/ipc4/handler.c:971: warning: unused variable ‘data’
CMakeFiles/sof.dir/build.make:379: recipe for target 'CMakeFiles/sof.dir/src/ipc/ipc4/handler.c.o' failed
make[2]: *** [CMakeFiles/sof.dir/src/ipc/ipc4/handler.c.o] Error 2
make[2]: *** Waiting for unfinished jobs....
/media/rander/linux_data/sof/src/ipc/ipc4/helper.c: In function ‘ipc_pipeline_new’:
/media/rander/linux_data/sof/src/ipc/ipc4/helper.c:159: error: ‘struct ipc4_pipeline_create’ has no member named ‘header’
/media/rander/linux_data/sof/src/ipc/ipc4/helper.c:161: error: ‘struct ipc4_pipeline_create’ has no member named ‘header’
/media/rander/linux_data/sof/src/ipc/ipc4/helper.c:162: error: ‘struct ipc4_pipeline_create’ has no member named ‘header’
/media/rander/linux_data/sof/src/ipc/ipc4/helper.c:162: error: ‘struct ipc4_pipeline_create’ has no member named ‘header’
/media/rander/linux_data/sof/src/ipc/ipc4/helper.c: In function ‘ipc_comp_connect’:
/media/rander/linux_data/sof/src/ipc/ipc4/helper.c:263: error: ‘struct ipc4_module_bind_unbind’ has no member named ‘header’
/media/rander/linux_data/sof/src/ipc/ipc4/helper.c:263: error: ‘struct ipc4_module_bind_unbind’ has no member named ‘header’
/media/rander/linux_data/sof/src/ipc/ipc4/helper.c: In function ‘ipc_comp_disconnect’:
/media/rander/linux_data/sof/src/ipc/ipc4/helper.c:325: error: ‘struct ipc4_module_bind_unbind’ has no member named ‘header’
/media/rander/linux_data/sof/src/ipc/ipc4/helper.c:325: error: ‘struct ipc4_module_bind_unbind’ has no member named ‘header’
/media/rander/linux_data/sof/src/ipc/ipc4/helper.c: In function ‘ipc4_create_chain_dma’:
/media/rander/linux_data/sof/src/ipc/ipc4/helper.c:559: error: ‘struct ipc4_chain_dma’ has no member named ‘header’
/media/rander/linux_data/sof/src/ipc/ipc4/helper.c:562: error: ‘struct ipc4_chain_dma’ has no member named ‘header’
/media/rander/linux_data/sof/src/platform/intel/cavs/platform.c: In function ‘platform_boot_complete’:
/media/rander/linux_data/sof/src/platform/intel/cavs/platform.c:322: error: ‘struct ipc_cmd_hdr’ has no member named ‘dat’
/media/rander/linux_data/sof/src/platform/intel/cavs/platform.c:323: error: ‘struct ipc_cmd_hdr’ has no member named ‘dat’
CMakeFiles/sof.dir/build.make:187: recipe for target 'CMakeFiles/sof.dir/src/platform/intel/cavs/platform.c.o' failed

**Comment by RanderWang** [comment]
This is my change for your reference.  "ipc_cmd_hdr" change will affect IPC3 build
````c
diff --git a/src/drivers/intel/cavs/ipc.c b/src/drivers/intel/cavs/ipc.c
index f29116de8..f87512306 100644
--- a/src/drivers/intel/cavs/ipc.c
+++ b/src/drivers/intel/cavs/ipc.c
@@ -243,8 +243,8 @@ int ipc_platform_send_msg(const struct ipc_msg *msg)
        ipc_write(IPC_DIPCIE, hdr->dat[1]);
        ipc_write(IPC_DIPCI, IPC_DIPCI_BUSY | hdr->dat[0]);
 #else
-       ipc_write(IPC_DIPCIDD, hdr->dat[1]);
-       ipc_write(IPC_DIPCIDR, IPC_DIPCIDR_BUSY | hdr->dat[0]);
+       ipc_write(IPC_DIPCIDD, hdr->ext);
+       ipc_write(IPC_DIPCIDR, IPC_DIPCIDR_BUSY | hdr->pri);
 #endif
 
        return 0;
diff --git a/src/ipc/ipc4/handler.c b/src/ipc/ipc4/handler.c
index a433ce468..9e08abb10 100644
--- a/src/ipc/ipc4/handler.c
+++ b/src/ipc/ipc4/handler.c
@@ -67,9 +67,9 @@ static int ipc4_delete_pipeline(struct ipc4_message_request *ipc4)
        struct ipc *ipc = ipc_get();
 
        pipe = (struct ipc4_pipeline_delete *)ipc4;
-       tr_dbg(&ipc_tr, "ipc4 delete pipeline %x:", (uint32_t)pipe->header.r.instance_id);
+       tr_dbg(&ipc_tr, "ipc4 delete pipeline %x:", (uint32_t)pipe->primary.r.instance_id);
 
-       return ipc_pipeline_free(ipc, pipe->header.r.instance_id);
+       return ipc_pipeline_free(ipc, pipe->primary.r.instance_id);
 }
 
 static int ipc4_comp_params(struct comp_dev *current,
@@ -967,29 +967,28 @@ void ipc_msg_reply(struct sof_ipc_reply *reply)
 
 void ipc_cmd(struct ipc_cmd_hdr *_hdr)
 {
-       struct ipc4_message_request *in = ipc_from_hdr(_hdr);
-       uint32_t *data = ipc_get()->comp_data;
+       struct ipc4_message_request in;
        enum ipc4_message_target target;
        int err;
 
-       if (!in)
-               return;
-
+       in.primary.dat = _hdr->pri;
+       in.extension.dat = _hdr->ext;
+       
        /* no process on scheduled thread */
        msg_data.delayed_reply = 0;
        msg_data.delayed_error = 0;
        msg_reply.tx_size = 0;
-       msg_reply.header = in->primary.dat;
-       msg_reply.extension = in->extension.dat;
+       msg_reply.header = in.primary.dat;
+       msg_reply.extension = in.extension.dat;
 
-       target = in->primary.r.msg_tgt;
+       target = in.primary.r.msg_tgt;
 
        switch (target) {
        case SOF_IPC4_MESSAGE_TARGET_FW_GEN_MSG:
-               err = ipc4_process_glb_message(in);
+               err = ipc4_process_glb_message(&in);
                break;
        case SOF_IPC4_MESSAGE_TARGET_MODULE_MSG:
-               err = ipc4_process_module_message(in);
+               err = ipc4_process_module_message(&in);
                break;
        default:
                /* should not reach here as we only have 2 message types */
@@ -1001,7 +1000,7 @@ void ipc_cmd(struct ipc_cmd_hdr *_hdr)
                tr_err(&ipc_tr, "ipc4: %d failed err %d", target, err);
 
        /* FW sends a ipc message to host if request bit is set*/
-       if (in->primary.r.rsp == SOF_IPC4_MESSAGE_DIR_MSG_REQUEST) {
+       if (in.primary.r.rsp == SOF_IPC4_MESSAGE_DIR_MSG_REQUEST) {
                char *data = ipc_get()->comp_data;
                struct ipc4_message_reply reply;
 
@@ -1012,8 +1011,8 @@ void ipc_cmd(struct ipc_cmd_hdr *_hdr)
 
                /* copy contents of message received */
                reply.primary.r.rsp = SOF_IPC4_MESSAGE_DIR_MSG_REPLY;
-               reply.primary.r.msg_tgt = in->primary.r.msg_tgt;
-               reply.primary.r.type = in->primary.r.type;
+               reply.primary.r.msg_tgt = in.primary.r.msg_tgt;
+               reply.primary.r.type = in.primary.r.type;
                if (msg_data.delayed_error)
                        reply.primary.r.status = msg_data.delayed_error;
                else
diff --git a/src/ipc/ipc4/helper.c b/src/ipc/ipc4/helper.c
index 271e16bf0..66ed5c4cc 100644
--- a/src/ipc/ipc4/helper.c
+++ b/src/ipc/ipc4/helper.c
@@ -156,10 +156,10 @@ int ipc_pipeline_new(struct ipc *ipc, ipc_pipe_new *_pipe_desc)
 {
        struct ipc4_pipeline_create *pipe_desc = ipc_from_pipe_new(_pipe_desc);
 
-       tr_dbg(&ipc_tr, "ipc: pipeline id = %u", (uint32_t)pipe_desc->header.r.instance_id);
+       tr_dbg(&ipc_tr, "ipc: pipeline id = %u", (uint32_t)pipe_desc->primary.r.instance_id);
 
-       return ipc4_create_pipeline(ipc, pipe_desc->header.r.instance_id,
-               pipe_desc->header.r.ppl_priority, pipe_desc->header.r.ppl_mem_size);
+       return ipc4_create_pipeline(ipc, pipe_desc->primary.r.instance_id,
+               pipe_desc->primary.r.ppl_priority, pipe_desc->primary.r.ppl_mem_size);
 }
 
 static int ipc_pipeline_module_free(uint32_t pipeline_id)
@@ -260,7 +260,7 @@ int ipc_comp_connect(struct ipc *ipc, ipc_pipe_comp_connect *_connect)
        int ret;
 
        bu = (struct ipc4_module_bind_unbind *)_connect;
-       src_id = IPC4_COMP_ID(bu->header.r.module_id, bu->header.r.instance_id);
+       src_id = IPC4_COMP_ID(bu->primary.r.module_id, bu->primary.r.instance_id);
        sink_id = IPC4_COMP_ID(bu->extension.r.dst_module_id, bu->extension.r.dst_instance_id);
        source = ipc4_get_comp_dev(src_id);
        sink = ipc4_get_comp_dev(sink_id);
@@ -322,7 +322,7 @@ int ipc_comp_disconnect(struct ipc *ipc, ipc_pipe_comp_connect *_connect)
        int ret;
 
        bu = (struct ipc4_module_bind_unbind *)_connect;
-       src_id = IPC4_COMP_ID(bu->header.r.module_id, bu->header.r.instance_id);
+       src_id = IPC4_COMP_ID(bu->primary.r.module_id, bu->primary.r.instance_id);
        sink_id = IPC4_COMP_ID(bu->extension.r.dst_module_id, bu->extension.r.dst_instance_id);
        src = ipc4_get_comp_dev(src_id);
        sink = ipc4_get_comp_dev(sink_id);
@@ -556,10 +556,10 @@ int ipc4_create_chain_dma(struct ipc *ipc, struct ipc4_chain_dma *cdma)
        uint32_t dir, host_chan, link_chan;
        int ret;
 
-       if (process_dma_index(cdma->header.r.host_dma_id, &dir, &host_chan) < 0)
+       if (process_dma_index(cdma->primary.r.host_dma_id, &dir, &host_chan) < 0)
                return IPC4_INVALID_NODE_ID;
 
-       if (process_dma_index(cdma->header.r.link_dma_id, &dir, &link_chan) < 0)
+       if (process_dma_index(cdma->primary.r.link_dma_id, &dir, &link_chan) < 0)
                return IPC4_INVALID_NODE_ID;
 
        /* build a pipeline id based on dma id */
diff --git a/src/platform/intel/cavs/platform.c b/src/platform/intel/cavs/platform.c
index 9053f7293..0160ac5f9 100644
--- a/src/platform/intel/cavs/platform.c
+++ b/src/platform/intel/cavs/platform.c
@@ -319,8 +319,8 @@ int platform_boot_complete(uint32_t boot_message)
        ipc_write(IPC_DIPCIE, header.dat[1]);
        ipc_write(IPC_DIPCI, IPC_DIPCI_BUSY | header.dat[0]);
 #else
-       ipc_write(IPC_DIPCIDD, header.dat[1]);
-       ipc_write(IPC_DIPCIDR, IPC_DIPCIDR_BUSY | header.dat[0]);
+       ipc_write(IPC_DIPCIDD, header.pri);
+       ipc_write(IPC_DIPCIDR, IPC_DIPCIDR_BUSY | header.ext);
 #endif
        return 0;
 }

File: .github/workflows/installer.yml (hunk 7 above)
Comment (mwasko) [reply .github/workflows/installer.yml]: > This removes IPC4 compilation for all platforms without replacing them with any IPC4 platform. So this is effectively turning off IPC4 compilation entirely which makes no sense and makes even less sense in an IPC4 pull request. This looks like trying to hide compilation issues.

@marc-hb I think that you are missing the point of this PR. The current IPC4 implementation has major design flaws that require code refactor. It is required to move forward and prepare us for Intel MTL platform support that officially will be released with IPC4 protocol. None of the cAVS targets has planned date to switch to support IPC4 protocol and all the cAVS releases are currently going with IPC3 - this is not a priority for now and cannot be blocking for MTL release.

We can check if there are some quick w/a that can be applied to sustain cAVS compilation with IPC4 for now, but it will be replaced with MTL IPC4 CI as soon as MTL support will be upstream. The cAVS IPC4 configuration support will require adaptation to incoming changes and depending on available bandwidth might require to be temporary turned off. 

The bottom line is that we cannot put MTL SOF + Zephyr IPC4 release at risk in favor of maintaining cAVS IPC4 support. If you have any problem with that then please contact me directly.

Comment (rfredzim) [reply .github/workflows/installer.yml]: @marc-hb I indeed introduced some errors in Friday's PR cleanup attempt and now fixed them.
I assure you I'm not attempting to hide compilation issues, but to cleanup ipc4 usage.
It's a bit disappointing to me that you read it this way, hopefully it won't discourage others from improving this FW.

From information I have cavs + ipc4 is not a valid configuration.
MTL + IPC4 is a valid one but I accept Liam's comment about MTL not being available to developers so in latest version I'm providing changes which should satisfy IPC4 + cavs config for the time being.

Please let me know if you have further comments.

**Review by lgirdwood — APPROVED** [review]

File: .github/workflows/installer.yml (hunk 7 above)
Comment (marc-hb) [reply .github/workflows/installer.yml]: > The current IPC4 implementation has major design flaws that require code refactor. [...]
> The bottom line is that we cannot put MTL SOF + Zephyr IPC4 release at risk in favor of maintaining cAVS IPC4 support.

I think any refactor that breaks cAVS IPC4 is fine as long as it compiles IPC4 code in CI "somewhere" else instead[*]. Any configuration can do. Merging code that is never compiled in CI does not move anything forward, it regresses and slows down everyone, breaks git bisect etc. If a huge jump is required then it can be submitted as a single giant commit.

Maybe I'm missing something here but... are we really discussing whether every git commit should compile?


[*] Other people actively developing IPC4 on IPC3 platforms may disagree.
